python/paddle/fluid/contrib/int8_inference/utility.py

```diff
@@ -0,0 +1,731 @@
+
+            if i.name in self._weights_var_name:
+                max_value = []
+                data = sampling_data[i.name][0]
+                for j in range(data.shape[0]):
+                    if not self._is_close(float(np.max(np.abs(data[j]))), 0.0):
+                        max_value.append(Calibrator.s8_max /
+                                         float(np.max(np.abs(data[j]))))
+                    else:
+                        max_value.append(0.0)
+                max_range = Calibrator.s8_max
+            else:
+                if i.name in self._conv_output_var_name:
+                    cur_op = self.program.current_block().ops[
+                        self._conv_output_var_name.index(i.name) + 2]
+                    if cur_op.has_attr('fuse_relu') and cur_op.attr(
+                            'fuse_relu'):
+                        max_range = Calibrator.u8_max
+                        self._u8_output_var.append(i.name)
+                    else:
+                        max_range = Calibrator.s8_max
+                        self._s8_output_var.append(i.name)
+                else:
+                    max_range = self.__get_max_range_by_var_name(self.program,
+                                                                 i.name)
+
+                max_value = [[np.abs(np_data)]
+                             for np_data in sampling_data[i.name]]
+
+            self._var_max_range[i.name] = max_range
+
+            self._var_max_value_map[i.name] = max_value
+
+    def __check_force_fp32_attr_by_output_var(self, program, var_name):
+        for op in program.current_block().ops:
+            if op.type == "conv2d" and var_name in op.output_arg_names:
+                return op.attr("force_fp32_output")
+        return False
+
+    def __get_op_index_by_output_var(self, program, var_name, start_index=0):
+        '''
+        Check the specified input variable is the output of the conv/pool2d
+        op's output or not.
+
+        Returns:
+            The index if the variable is the output of any conv/pool2d op's
+            output.
+            -1 when the variable is not the output of any conv/pool2d op's 
+            output.
+        '''
+        for index, op in enumerate(program.current_block().ops[start_index:]):
+            if var_name in op.output_arg_names and op.type in Calibrator.supported_int8_op_type:
+                return index
+        return -1
+
+    def __get_op_index_by_input_var(self, program, var_name, start_index=0):
+        '''
+        Get the op index by specified input variable.
+        Returns:
+            The op index if the variable is the input of this op or -1 if the 
+            variable is not the input of any op. 
+        '''
+        for index, op in enumerate(program.current_block().ops[start_index:]):
+            if var_name in op.input_arg_names:
+                return index
+
+        return -1
+
+    def __get_quantize_dequantize_combination(self, program):
```

Comment:
Could this function be made smaller? It's very complex, I'm having hard time analyzing this.